core/src/main/java/com/oath/oak/InternalOakMap.java

```diff
@@ -434,42 +410,51 @@ V put(K key, V value, Function<ByteBuffer, V> transformer) {
         if (ei == -1) {
             ei = c.allocateEntryAndKey(key);
             if (ei == -1) {
-                return rebalancePutIfAbsent(c, key, value, transformer);
+                rebalance(c);
+                return putIfAbsent(key, value, transformer);
             }
             int prevEi = c.linkEntry(ei, true, key);
             if (prevEi != ei) {
-                Handle handle = c.getHandle(prevEi);
-                if (handle == null) {
-                    ei = prevEi;
-                    prevHi = c.getHandleIndex(prevEi);
-                } else {
+
+                prevHi = c.getHandleIndex(prevEi);
+                if (prevHi != -1 ) {
```

Comment:
wait - so if you are changing this code does this mean it no longer follows the pseudocode in Hagar's thesis? so how do we know the proofs apply here? there is no straightforward way to "see" the code is correct

Reply:
This code is actually closer to Hagars thesis.